clustermesh: correctly release allocator references for GC

oblazek opened the pull request:
On each reconnect we failed to release all references to the old Allocator, meaning it was not GCed properly, leaving out some parts still in memory -> resulting in a memleak.

By launching NewAllocator (a remote one) with `WithoutGC()` option we save a goroutine which is noop anyway (for the remote connection) and which was preventing the remoteAlloc to be released.

This commit fixes it.

Parca memory profile comparisons:
 
Master (before https://github.com/cilium/cilium/pull/23532 got merged) :
![master1](https://user-images.githubusercontent.com/15906154/219064739-30731ff2-f744-4dee-91eb-04f7d0712848.png)

With https://github.com/cilium/cilium/pull/23532 fixes:
![PR-23532](https://user-images.githubusercontent.com/15906154/219061877-1ea61446-8a87-4c19-a7bb-7e28bd364b30.png)

After this fix:
![fixed](https://user-images.githubusercontent.com/15906154/219061913-4dc90d8c-115b-42d1-b324-bf2b2eebc371.png)

Fixes: #22988
Fixes: #13446

```release-note
Fix memory leak caused on clustermesh reconnect.
```

Signed-off-by: Ondrej Blazek <ondrej.blazek@firma.seznam.cz>

aditighag commented on pkg/allocator/allocator.go:
```diff
@@ -389,10 +389,6 @@ func (a *Allocator) GetEvents() AllocatorEventChan {
 func (a *Allocator) Delete() {
 	close(a.stopGC)
 	a.mainCache.stop()
-
-	if a.events != nil {
-		close(a.events)
-	}
```
This should've been cleaned up on L266 anyway in the `Close()` call so why move it outside of the `Delete()` function? Please mention the reasoning in the commit description.

oblazek replied:
np,  so imho the right thing to do when closing the remoteCache would be to delete the whole remoteAllocator - i.e. running `remoteAlloc.Delete()` -- but the problem with that is that without moving this code outside it would close the event channel not owned by the remoteAlloc, but owned by the main `CachingIdentityAllocator`. This in fact breaks the event channel which is used to pass remoteIdentity events to the main allocator (and panics on writing to closed channel).

aditighag commented on pkg/identity/cache/allocator.go:
```diff
@@ -449,7 +453,7 @@ func (m *CachingIdentityAllocator) WatchRemoteIdentities(remoteName string, back
 	}
 
-	remoteAlloc, err := allocator.NewAllocator(&key.GlobalIdentity{}, remoteAllocatorBackend, allocator.WithEvents(m.IdentityAllocator.GetEvents()))
+	remoteAlloc, err := allocator.NewAllocator(&key.GlobalIdentity{}, remoteAllocatorBackend, allocator.WithEvents(m.IdentityAllocator.GetEvents()), allocator.WithoutGC())
```
What's the rationale of calling an allocator without GC? (Unfortunately, the existing code is super light on explanation for the GC part.)

On a related note, if an allocator is allocated without GC, then what does the stopping of the gc channel do in the `Delete()` - 

```
func (a *Allocator) Delete() {
	close(a.stopGC)
```

oblazek replied:
if the GC is enabled.. the code starts this go routine:
```
if !a.disableGC {
		go func() {
			select {
			case <-a.initialListDone:
			case <-time.After(option.Config.AllocatorListTimeout):
				log.Fatalf("Timeout while waiting for initial allocator state")
			}
			a.startLocalKeySync()
		}()
	}
```
which in fact starts this go routine:
```
go func(a *Allocator) {
		kvTimer, kvTimerDone := inctimer.New()
		defer kvTimerDone()
		for {
			if err := a.syncLocalKeys(); err != nil {
				log.WithError(err).Warning("Unable to run local key sync routine")
			}

			select {
			case <-a.stopGC:
				log.Debug("Stopped master key sync routine")
				return
			case <-kvTimer.After(option.Config.KVstorePeriodicSync):
			}
		}
	}(a)
```
if I do not stop the `stopGC` channel I have no way of stopping the goroutine which is the problematic part.

So the GC (and these goroutines) only makes sense for the local allocator (as they sync the local keys to the backend) but only for that (as there are no local keys for the remote backend).

aditighag reviewed (commented):
Great find! Couple of inline comments on the fix. 

The graphs are good as a before and after comparison, but they don't point to the objects leading to memory leak. Any chance you can attach a snippet of heap dump?
^^I echo Casey's comment.

This isn't a blocking comment for your PR, but I encourage code reviewers more familiar with this code path to think about unit/control/integration tests to proactively catch such regressions.

giorio94 commented on pkg/identity/cache/allocator.go:
```diff
@@ -264,6 +264,10 @@ func (m *CachingIdentityAllocator) Close() {
 	}
 
 	m.IdentityAllocator.Delete()
+	if m.events != nil {
+		close(m.events)
+	}
```
There is a potential issue here. The events channel gets initialized in the [`NewCachingIdentityAllocator` function](https://github.com/cilium/cilium/blob/master/pkg/identity/cache/allocator.go#L230-L241). According to the comments of the `Close` function, once it is executed, it should be possible to call `InitIdentityAllocator` again, which though uses the events channel. Hence, it would break, given that the channel would be closed.

~~It doesn't seem to me that `Close` is ever used outside of tests, so no actual difference should be noticed, still, I feel it might be better not to close that channel to better comply with the desired approach.~~

Edit: I missed [this usage](https://github.com/cilium/cilium/blob/8aa8dd5702f93b480d252702df632cd934650664/daemon/cmd/daemon.go#L1400); I adapted the comment based on that.

giorio94 replied:
I don't see any particularly clean ways for handling this (at the moment it does not lead to issues, since the allocator is never initialized again after having been closed). I wonder if the other code reviewers have any suggestion.

oblazek replied:
yeah you are right about the comment, hmm.. if anybody sees a potential problem with this I can just revert this. The oneliner that fixes the memleak here is sufficient as well (`allocator.WithoutGC()`)

squeed replied:
I think the code comment is moot; I see no evidence of Close() being called except for on daemon shutdown. I think we can just remove that comment and, if desired, panic on `InitIdentityAllocator` being called twice.

The real question: would this change be safe to backport? I'm checking.

squeed replied:
In versions v1.13, v1.12, and v1.11, Close() isn't even called. So it's safe to say this comment is not useful.

aanm reviewed (commented):
I'll defer my review to @giorio94

squeed reviewed (changes requested):
Just need to change the docblock in CachingIdentityAllocator.Close() and this is good to go.

oblazek committed:
allocator: fully close remote cache on connection restart

Previously when remote identity cache was closed it did not stop
the stopGC channel, therefore it leaked the allocator resources.

This commit fixes the issue and properly deletes all resources and
closes the stopGC channel, but closing of events channel is moved
outside of `allocator.Delete()` func as the channel must be closed
only by the main CachingIdentityAllocator (it's shared by each
remote allocator, but not owned).

Signed-off-by: Ondrej Blazek <ondrej.blazek@firma.seznam.cz>

oblazek committed:
allocator: start remote allocators withoutGC option

Previously each remote allocator spawned a goroutine which
is supposed to sync all local keys to the remote etcd. But
this makes sense only for the main identityAllocator, not
for remote one(s).

This commit fixes it, uses a disableGC option to save the
goroutine.

Signed-off-by: Ondrej Blazek <ondrej.blazek@firma.seznam.cz>

giorio94 reviewed (approved):
/lgtm Thanks @oblazek

aditighag reviewed (approved):
Thanks! Looks good for my code owners.
